### PR TITLE
feat(NcButton): Add `pressed` state for stateful buttons

### PR DESCRIPTION
**Pull request by susnux**
### ☑️ Resolves

* Alternative to #4280
* Resolves https://github.com/nextcloud/server/issues/37109

This simply adds a pressed state and enforces `primary` on pressed, and any other type (except primary) otherwise.
Not we should probably remove the button variant of the checkboxradioswitch for not grouped elements to prevent different styled components with the same function.

### 🖼️ Screenshots

https://github.com/nextcloud/nextcloud-vue/assets/1855448/b6716849-64ce-4bfb-93ac-1c28a0302db5

### 🏁 Checklist

- [x] ⛑️ Tests are included or are not applicable
- [x] 📘 Component documentation has been extended, updated or is not applicable

### CODE REVIEW

File: src/components/NcButton/NcButton.vue [hunk 1]
```diff
@@ -196,6 +196,66 @@ button {
...
+	computed: {
+		ariaLabel() {
+			return this.isFavorite ? 'Remove as favorite' : 'Add as favorite'
+		},
+	},
+	methods: {
+		toggleFavorite() {
+			this.isFavorite = !this.isFavorite
+		},
+	},
```
Comment (ShGKme) [inline src/components/NcButton/NcButton.vue:255]: What about adding `update:pressed` event to the component?
Then `pressed` state will be synced and component's user won't need to handle clicks manually.

File: src/components/NcButton/NcButton.vue [hunk 2]
```diff
@@ -352,6 +440,15 @@ export default {
...
 					...this.$listeners,
 					click: ($event) => {
+						// Update pressed prop on click if it is set
+						if (typeof this.$props.pressed === 'boolean') {
+							/**
+							 * Update the current pressed state of the button (if the `pressed` property was configured)
+							 *
+							 * @property {boolean} newValue The new `pressed`-state
+							 */
+							this.$emit('update:pressed', !this.$props.pressed)
```
Comment (ShGKme) [inline src/components/NcButton/NcButton.vue:450]: Let's also list the event in the `emits` option

File: src/components/NcButton/NcButton.vue [hunk 3]
```diff
@@ -352,6 +440,15 @@ export default {
 				on: {
 					...this.$listeners,
 					click: ($event) => {
+						// Update pressed prop on click if it is set
+						if (typeof this.$props.pressed === 'boolean') {
```
Comment (ShGKme) [inline src/components/NcButton/NcButton.vue:444]:
```suggestion
						if (typeof this.pressed === 'boolean') {
```

File: src/components/NcButton/NcButton.vue (hunk 2 above)
Comment (ShGKme) [inline src/components/NcButton/NcButton.vue:450]:
```suggestion
							this.$emit('update:pressed', !this.pressed)
```

File: src/components/NcButton/NcButton.vue (hunk 3 above)
Comment (ShGKme) [reply src/components/NcButton/NcButton.vue:444]: It seems to me, we always use shorthand props access

Comment (susnux) [reply src/components/NcButton/NcButton.vue:444]: ok

**Review by marcoambrosini — APPROVED** [review]
Nice work, thank you :)

**Review by JuliaKirschenheuter — APPROVED** [review]
haven't tested but looks good

**Review by ShGKme — APPROVED** [review]
Tested, works fine, looks good by code

**Comment by ShGKme** [comment]
Cypress tests seems to be related

**Review by jancborchardt — APPROVED** [review]
Nice, looks good! :)

**Comment by susnux** [comment]
> Cypress tests seems to be related

Yes as this is a visual change, the snapshots need to be updated, but the other tests pass :)

**Comment by susnux** [comment]
Oh yes, I love that cypress on my system renders the snapshots differently than on CI (antialiasing)...

But now the tests are still flaky. As the cursor within the input "flashes" and that is causing the snapshots to differ (compare runner 1 and runner 2)

image

**Pull request merged**